OpenRA.Mods.Common/Graphics/ContrailRenderable.cs

```diff
@@ -73,11 +73,31 @@ public void Render(WorldRenderer wr)
 			// Start of the first line segment is the tail of the list - don't smooth it.
 			var curPos = trail[Index(next - skip - 1)];
 			var curColor = color;
-			for (var i = 0; i < length - skip - 4; i++)
+
+			for (var i = 1; i < renderLength; i++)
 			{
-				var j = next - skip - i - 2;
-				var nextPos = Average(trail[Index(j)], trail[Index(j - 1)], trail[Index(j - 2)], trail[Index(j - 3)]);
-				var nextColor = Exts.ColorLerp(i * 1f / (length - 4), color, Color.Transparent);
+				var j = next - skip - 1 - i;
+				WPos nextPos;
+				int maxSmoothLength = 4; // Use 4 contrail position in circular buffer for max smooth effect.
+
+				var alpha = (renderLength - i) * color.A / renderLength;
+				if (alpha == 0)
+					return;
+
+				var nextColor = Color.FromArgb(alpha, color); // Use bits shift to avoid big rounding errors.
```

Comment:
You still haven't addressed my request about this comment.

Reply:
Removed